dali/kernels/test/warp_test/warp_cpu_test.cc

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <opencv2/imgcodecs.hpp>
+#include <opencv2/imgproc.hpp>
+#include <string>
+#include <vector>
+#include "dali/kernels/imgproc/warp_cpu.h"
+#include "dali/kernels/imgproc/warp/affine.h"
+#include "dali/kernels/test/tensor_test_utils.h"
+#include "dali/kernels/test/dump_diff.h"
+#include "dali/kernels/test/mat2tensor.h"
+#include "dali/kernels/test/test_tensors.h"
+#include "dali/kernels/scratch.h"
+#include "dali/kernels/alloc.h"
+#include "dali/test/dali_test_config.h"
+#include "dali/core/geom/transform.h"
+#include "dali/kernels/test/warp_test/warp_test_helper.h"
+
+namespace dali {
+namespace kernels {
+
+inline void IsWarpKernelValid() {
+  check_kernel<WarpCPU<AffineMapping2D, 2, float, uint8_t, float>>();
+}
+
+TEST(WarpCPU, Affine_Transpose_Single) {
+  AffineMapping2D mapping_cpu = mat2x3{{
+    { 0, 1, 0 },
+    { 1, 0, 0 }
+  }};
+
+  cv::Mat cv_img = cv::imread(testing::dali_extra_path() + "/db/imgproc/alley.png");
+  auto cpu_img = view_as_tensor<uint8_t>(cv_img);
+
+  WarpCPU<AffineMapping2D, 2, uint8_t, uint8_t, BorderClamp> warp;
+
+  ScratchpadAllocator scratch_alloc;
+
```

Comment:
I got confused here. On one hand you use `Scratchpad*` here, on the other hand it doesn't seem to be used in `WarpCPU` class.

Reply:
Handling scratchpad memory is part of the Kernel API contract. I know it's not used _now_, but kernels should _never_ be used without a valid scratchpad - otherwise the program is ill-formed.

Reply:
> kernels should never be used without a valid scratchpad

So, there's nothing about this requirement in the blueprints, so I assume it's not a requirement. Test not only verifies the code, but also documents it: creates an example of usage. With this here used gets confused.

So, IMHO, either let's enforce using `Scratchpad` in API (like the `check_kernel` or sth) or remove it from here. Just to be consistent

Reply:
From the documentation of Run in `kernel.h`:
```
   * @param context - environment; provides scratch memory, cuda stream, batch info, etc.
   *                  Scratch area must satisfy requirements returned by Setup.
```